Feature/arbitrary tupleset

agentm opened the pull request:
I created this pull request from your branch so that I could review the code in context. Feel free to create pull requests, too.

agentm commented on src/bin/TutorialD/tutd.hs:
```diff
@@ -82,6 +82,9 @@ printWelcome = do
 
 main :: IO ()
 main = do
+  hSetEncoding stdin  utf8
```
What issue does this fix? Is this a platform-specific fix? I have been able to print out UTF-8 characters without this.

YuMingLiao replied:
Thanks for the review.

Yes, it's a platform-specific fix. I use nixos in virtualbox and stack. And my stack don't parse locale settings. It turns en-US.utf8 into posix. If I don't add these lines, TutorialD command will fail. Haven't figure that out.

YuMingLiao replied:
I will take that off when I pull request. Sorry for the confusion.

YuMingLiao commented on src/lib/ProjectM36/Base.hs:
```diff
@@ -527,3 +538,75 @@ instance Hashable DatabaseContextFunction where
+      V.fromList <$> vectorOf maxbound arbitrary
+
+--not sure what's use on this one
+instance {-# OVERLAPPING #-} Arbitrary TypeVarMap where
```
Thanks for the comment! It's encouraging! And, yes. -30 to 30 is the default range of an arbitrary integer. Should I unleash this? I don't have clear image of what an arbitrary feature should be in my mind. 

I guess fixed `Attribtue`s is kind of limiting. Let me try to implement all of them this time.

YuMingLiao commented on src/lib/ProjectM36/Base.hs:
```diff
@@ -527,3 +538,75 @@ instance Hashable DatabaseContextFunction where
+arbitrary' TextAtomType = TextAtom <$> (arbitrary :: Gen Text)
+arbitrary' DayAtomType  = DayAtom <$> (arbitrary :: Gen Day)
+arbitrary' BoolAtomType  = BoolAtom <$> (arbitrary :: Gen Bool)
+arbitrary' _            = undefined
```
Yes, I implement simple atomtypes for a start, only to get some feedback. I have trouble with imaging the usage of other atomtypes. I don't know if arbitrary Constructed Atomtypes is needed since it is user-defined. When I random things, it feels I didn't offer enough meaningful data. But I guess this arbitrary feature is for testing some operation functionality, right? Sub-relation is a good point, thanks for reminding me.

I'll look into sub-relations and construct atomtypes, making it complete.
and do the cleanup you mentioned. 

Thanks again for the comment and review.

YuMingLiao replied:
arbitrary IntAtom is added now. 

However, I noticed,
TutorialD (master/main): x:=relation{name Text, age Int}{tuple{name "Steve", age 40},tuple{name "Mike", age 31}}
ERR: AtomTypeMismatchError IntAtomType IntegerAtomType

And there is no intP for atomP in code. So there seems no input way for IntAtom.

reference: issue [#138](https://github.com/agentm/project-m36/issues/138)

agentm replied:
I think that a version one of this patch need not necessarily include arbitrary attributes. But, if you think it's easy-to-add, that's fine.

Regarding the `Int` mismatch, the issue is that the default type for `40` is Integer. Project:M36 doesn't support typeclasses yet, but it would be nice to infer the type from the attributes. I'll open a ticket for this specific issue.

agentm replied:
I added an `int()` function in #201.

Example: `:showexpr relation{a Int}{tuple{a int(4)}}`

YuMingLiao replied:
I was trying to create an arbitrary relation with sub-relation. So first, I need make an attribute list with an relation attribute in it.

TutorialD (master/main): a :: {name Text, age Integer}
TutorialD (master/main): createarbitraryrelation b {name Text, age Integer, rel a} 1-2
TutorialD (master/main): :showexpr b
tutd: Prelude.undefined
It is becaue I haven't defined `TypeVariableType`
`arbitrary' (TypeVariableType tvname) = undefined`

How could I define an relation with sub-relation?

YuMingLiao replied:
To make arbitrary sub-relation occur, then a way to define its attributes in `DEFINE ::` or `ASSIGN :=` is need.
Something like `a :: {name Text, rel relation{name Text, age Integer} }` is needed.
I haven't figure out the whole expression yet.

And I don't understand these two things, ConstructedAtomType and TypeVariable, for now. So I don't know when an arbitrary instance of these two will be used. 

I don't think arbitrary attributes is necessary either.

YuMingLiao replied:
I notice that an arbitrary subrel can be created by `group` after `createarbitraryrelation`. 
```
createarbitraryrelation s {city Text,s# Integer,sname Text,status Integer} 1-2
showexpr s group ({sname,status,s#} as subrel)
```

So I guess the implementation is unncessary.

agentm replied:
The lack of ability to construct to create a relation type is just a bug. I'll work on fixing that in master so that you can proceed. Thanks for the report!

YuMingLiao replied:
Thanks! What a great project!

YuMingLiao commented on src/lib/ProjectM36/Base.hs:
```diff
@@ -527,3 +538,75 @@ instance Hashable DatabaseContextFunction where
+arbitrary' _            = undefined
+
+getAtomType :: Attribute -> AtomType
+getAtomType (Attribute _ atomType) = atomType
```
I see.

But if I import ProjectM36.Attribute in Base.hs, it cause a cycle of ...
module Error import Base
module Base import Attribute
module Attribute import Error

Haven't figure out relationships between them. not dare to move. So I'll keep that way first.

agentm commented:
I fixed #203, so the nested, arbitrary relation construction should be able to be coded at line 610 in Base.hs. @YuMingLiao, would you like to finish this work?

YuMingLiao commented:
Sure! I will finish it. I am working on arbitrary DateTime now. I may need some time to understand git mechanism, ConstructedAtomType and TypeVariable. 

Thanks for your support!

agentm commented:
It occurred to me that the `ConstructedAtomType` `Arbitrary` instance will require access to the `DatabaseContext` since the `AtomType` alone does not contain the list of available value constructors. If you want to pass that instance on to me to finish, that's fine.

YuMingLiao commented:
Good news! I have managed to create arbitrary `ConstructedAtomType`!

```
TutorialD (master/main): createarbitraryrelation b {a Hair} 5-10
TutorialD (master/main): :showexpr b
┌─────────────┐
│a::Hair      │
├─────────────┤
│White        │
│White        │
│Other "Sunny"│
│Other "Mary" │
│Black        │
│Other "Mary" │
│White        │
│Black        │
│Other "Ted"  │
└─────────────┘
```

But I still don't understand the usage of `TypeVariable`. Could you give me an example of that in TutorialD expression?

agentm commented:
By the time `arbitrary'` sees the `AtomType`, all types should be fully resolved, so you shouldn't see any `TypeVariable`s. I've been thinking about splitting `AtomType`s into resolved and unresolved types, but I would need to do some refactoring first.

Thanks for your hard work! I'll test out your branch today.

YuMingLiao commented:
You are welcome! Take your time, and thanks for the great project!

agentm commented:
I fixed the attribute resolution via #205, so there are new possibilities parsing. In testing, I found a case that results in an error from within QuickCheck:

```
TutorialD (master/main): createarbitraryrelation y {spam Int, spam2 Interval Int} 3-100
ERR: UnhandledExceptionError "QuickCheck.elements used with empty list\nCallStack (from HasCallStack):\n  error, called at ./Test/QuickCheck/Gen.hs:186:15 in QuickCheck-2.10.0.1-ALuGysu7txTB2VnkPyDuTq:Test.QuickCheck.Gen"
```
Could you take a look?

I also added some basic tests for this feature: `cabal test test-tutoriald`.

YuMingLiao commented:
It happens in `arbitray' ConstructedAtomType`,
I assume Interval should be parsed as `IntervalAtomType atomTy`, but it is parsed as a `ConsturctedAtomType`.

```
TutorialD (master/main): createarbitraryrelation a {spam Interval Int} 3-100
```
 result parsed by `makeAttributeExprsP`
```
[AttributeAndTypeNameExpr "spam" (ADTypeConstructor "Interval" [ADTypeConstructor "Int" []]) ()]
```
result while retrieving a `TypeConstructor` in  `arbitray' ConstructedAtomType`

```
Just (ADTypeConstructorDef "Interval" ["a"],[])
```

I expected the `TypeConstructor` found at least have one `DataConstructorDef` (Am my assumption right? data TypeConstructorName = DataConstructorDef1 ...), but Interval has none.
Hence while the program randomly choose a `DataConstructorDef` by elements, it cause this empty list error.

Why is it not parsed as `IntervalAtomType`? Is it deprecated?
What is the difference between ADTypeConstructor and PrimitiveTypeConstructor?

I'll look into it more tomorrow.

YuMingLiao commented:
I see now, in test, you expect Interval to be `(ConstructedAtomType "Interval" (M.singleton "a" IntegerAtomType)`

agentm commented:
I see- this may be another primitive AtomType bug. I'll take a closer look.

YuMingLiao commented:
Okay, I see that now.

```
basicTypeConstructorMapping :: TypeConstructorMapping
basicTypeConstructorMapping = primitiveTypeConstructorMapping ++
                              maybeTypeConstructorMapping ++
                              eitherTypeConstructorMapping ++
                              listTypeConstructorMapping ++
                              intervalTypeConstructorMapping

```
`intervalTypeConstructorMapping` is different from `primitiveTypeConstructorMapping`.

```
atomTypeForTypeConstructor tCons tConss tvMap = case findTypeConstructor (TC.name tCons) tConss of
  Nothing -> Left (NoSuchTypeConstructorError (TC.name tCons))
  Just (PrimitiveTypeConstructorDef _ aType, _) -> Right aType
  Just (tConsDef, _) -> do
          tConsArgTypes <- mapM (\tConsArg -> atomTypeForTypeConstructor tConsArg tConss tvMap) (TC.arguments tCons)
          let pVarNames = TCD.typeVars tConsDef
              tConsArgs = M.fromList (zip pVarNames tConsArgTypes)
          Right (ConstructedAtomType (TC.name tCons) tConsArgs)
```
So `atomTypeForTypeConstructor` thinks it is a `ConstructedAtomType`.

I noticed that primitiveTypeConstructorP is taken off. Now all the primitive types become a `ADTypeConstructor` first, then turned back when `findTypeConstructor`.
I guess that's what you said eariler about more possibility parsing.

So, should Interval be a `primitiveType`? Or it should be an `ConstructedAtomType`?
like data Interval a = Interval a a Bool Bool ?

agentm commented:
`Interval` should ultimately become a `ConstructedAtomType` since it is clearly composed of primitives. I chose not to implement it that way because of some special syntax around it. For example, `Interval` does not have any visible data constructors and it prints and gets parsed (in CSV) specially. Since this current design confused me, I am working on making it a `ConstructedAtomType`.

YuMingLiao commented:
After I make changes in `atomTypeForTypeConstructor`. Interval can be generated now.
```
TutorialD (master/main): createarbitraryrelation e {spam Interval Day} 3-5
[AttributeAndTypeNameExpr "spam" (ADTypeConstructor "Interval" [ADTypeConstructor "Day" []]) ()]
TutorialD (master/main): :showexpr e
┌───────────────────────┐
│spam::Interval Day     │
├───────────────────────┤
│[1858-12-06,1858-12-06)│
│(1858-10-20,1858-10-20)│
│(1858-12-16,1858-12-16]│
│(1858-11-11,1858-11-11]│
└───────────────────────┘
```

But as you can see, there is some logic error in generating an Interval...haha

YuMingLiao commented:
```
TutorialD (master/main): createarbitraryrelation e {spam Interval Day} 3-5
TutorialD (master/main): :showexpr e
┌───────────────────────┐
│spam::Interval Day     │
├───────────────────────┤
│[1858-12-10,1858-10-20]│
│[1858-12-11,1858-11-18)│
│[1858-11-12,1858-12-05]│
└───────────────────────┘

```
it has two different values now.

Interval may need some restriction. The type in it can only be an instance of Comparable typeclass or something like that.

agentm commented:
Yea, we'll need some Interval-specific code to make the arbitrary instance generate valid intervals. In the future, users may have to specify their own arbitrary instances.

Indeed, I'm feeling more aware that we need to support typeclasses for custom types. Such classes would be useful for import/export, validation, etc.

I should have a replacement `Interval` type ready tomorrow.

YuMingLiao commented:
I see. Never mind. I'm just trying to understand more code.

Good job! Thanks.

Where should I look to fetch your version? Maybe I can try if Interval works as expected.

About typeclass, I feel the same way. If Atomtype for Interval can be `gt` or `lt`, Then it is not far from having a sorted `DataFrame`.

agentm commented:
The `Interval` changes are on [my arbitrary_tupleset branch](https://github.com/agentm/project-m36/tree/YuMingLiao-feature/arbitrary_tupleset). I can still reproduce the `Gen` error using `relation{a Interval Int}`. Which line in `Arbitrary` causes the failure?

YuMingLiao commented:
It's `elements` in `arbitrary (ConstructedType...`, about line 80.
it's because there is no DataConstructorDef for Interval. So `elements dcDef` choose from empty list and cause error .
If Interval is ADTypeConstructor, shouldn't it be defined as
`data Interval a = Interval a a Bool Bool`
so it has at least one `DataConstructor` ?
Am I misunderstood?
(time to bed...good day/night!)

agentm commented:
Ah, I understand. In this case, I chose not to include a public data constructor because that would allow construction of invalid intervals (such as `Interval 1 0`). Users must use the `AtomFunction` `interval` to construct `Interval`s which validates the values. This is similar to Haskell module exports/hiding.

Until we have typeclasses which would allow user-defined types to implement their own `Arbitrary` instances, I think I will add a special case for `Interval`s. I suspected a special case would be required anyway because dropping random values into an `Interval` is likely to create invalid values anyway.

What do you think? Is there a better approach?

agentm commented:
I added a special case to generate valid `Interval`s in `Arbitrary.hs`. Let me know what you think.

YuMingLiao commented:
Sorry I couldn't give instant feedback. I don't have strong opinion about this special Interval thing. It can wait.
But you implemented it so quick. And It works as expected! That's brilliant! Thanks!

```
TutorialD (master/main): createarbitraryrelation b {spam Interval Int} 3-5
TutorialD (master/main): :showexpr b
┌──────────────────────────┐
│spam::Interval (a::Int)   │
├──────────────────────────┤
│Interval 11 30 False True │
│Interval -27 16 False True│
│Interval -22 -9 True True │
│Interval -8 1 False False │
└──────────────────────────┘
```

And recently I found [fake](http://hackage.haskell.org/package/fake)
It can generate readable and plausible things like names, id, address.
Do you think it's worth implementing?

agentm commented:
`Fake` could be interesting for end-users, but the default types are generic. We could certainly offer additional types as useful add-ons. I think the additional types would be more useful with typeclasses, though.

Are there any more tests you can think of that would be useful?

YuMingLiao commented:
what do you mean by additional types? I thought ConstructedAtomType already serves additional type.

I was thinking something like ...
createarbitraryrelation a {name Text, id Int, address Text, phoneNumber Text} 3-5
Then it will create a more readable, plausible relation than what the arbitrary function can give now.

agentm commented:
I thought we could offer a `PhoneNumber` type which would have an `Arbitrary` instance to create phone numbers and so on.

YuMingLiao commented:
Oh, I see what you mean now. That makes sense. The additional type can have some more property related to reality like PhoneNumber format validation. 
and it can create arbitrary relation with additional types and let user make their own attribute names.
`createarbitraryrelation a {employeeName Name, companyID ID, phoneNumber PhoneNumber} 3-5`

So it is worth a try, right? easily creating more user case relation to play with? Let me dive into this...

agentm commented:
Add-on types would certainly be useful, but they don't need to part of this feature branch.

I was thinking that there could be a central repository for add-on types and schemas so that they could be easily loaded dynamically at runtime, but that's a future feature.

YuMingLiao commented:
Okay then, you are right. And it seems odd to add AddOn types to ConstructedAtomType. I guess I just want to create more relations to play with. :p ... So I guess this is it. I've committed the final work and push to my feature/arbitrarau_tupleset. I'll pull a request.

YuMingLiao commented:
Okay, I guess I don't need to pull new request because this request is already following my branch.

agentm commented:
I pulled from your branch but it no longer compiles. It looks like `arbitrary'` in `Arbitrary.hs` disappeared or was moved. Do you see the same?

YuMingLiao commented:
Oops.  I have git pushed again. This time should be fine. Sorry for the inconvenience.

agentm commented:
This is now merged. Thanks for your hard work on this feature- this is an unusual feature that other DBMSes don't have.

YuMingLiao commented:
Glad to finish my first open source contribution here. Thanks for the support, mentoring and encouragement.
